_posts/2015-11-25-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md

```diff
@@ -0,0 +1,197 @@
+- functional programming
+- dotnet
+- fsharp
+- fscheck
+- property based testing
+- visual studio
+---
+
+### Background
+Craftsmanship in software strives for high, predictable, repeatable and affordable quality; this is the reason why we care about design, code reviews and of course testing. 
+Test Driven Development helps us create a flexible implementation that meets the expectations of our clients in terms of features, performance and quality; as professionals we have to master as many tools as possible to reach our goal of steadily providing value. 
+	
+### Property based testing
+The above is the reason why I am exploring Property Based Testing; I like to think about it as *"Let the machine do the heavy lifting for you"* or as [John Huges] says *"Don't write tests, generate them"*.
+
+I am not claiming that you should stop writing unit tests, I find that when driving the design [Outside-in] hence declaring the relations between my components, unit tests are the tool to use.
+However I sometimes want to discover deeper properties of my domain, and sometimes I am faced with features that have more edge cases than my brain could handle. 
+In such cases, generating a wide array of inputs to test my program is appealing; this is where I think property testing becomes a viable tool.
+
+	
+If you are asking "How can I set up a development environment in .Net that would allow me to experiment and work with this tool?" the following tutorial is for you. 
+
+### Development environment details
+
+* Windows 10 Home Edition
+* Visual Studio 2013 version 12 Update 5
+* Resharper version 8.2.3
+* NCrunch version 1.48.0.5
+* FsCheck version 2.2.2
+* FsCheck.NUnit version 2.2.2
+* NUnit version 2.6.4
+
+### Steps
+1. Open Visual Studio and create an F# Project
+2. Open the Nuget Package Manager following *TOOLS > Nuget Package Manager > Package Manager Console* from the main menu and type
+    1. ```Install-Package FsCheck```
+    2. ```Install-Package FsCheck.Nunit```
+3. Comment out the content in **FsCheckAddin.fs** file to allow [NCrunch] to run the tests
+    
+	``` fs
+        //[<NUnitAddin(Description = "FsCheck addin")>] 
+		//type FsCheckAddin() = 
+		//  interface IAddin with 
+		//      override x.Install host = 
+		//          let tcBuilder = new FsCheckTestCaseBuilder() 
+		//          host.GetExtensionPoint("TestCaseBuilders").Install(tcBuilder) 
+        //          true 
+   ```
+   
+4. In your test file add the following
+    
+	``` fs
+        module Properties = 
+        
+            open NUnit.Framework 
+            open FsCheck
+    ```
+	
+5. Add this test to enable [NCrunch]
+    
+	``` fs
+            //Needed to enable NCrunch 
+            [<Ignore>][<Test>] 
+            let ignoreMe() = () 
+    ```
+	
+6. Create a static class whose methods are the properties you want to test. For this example I'll test a plainly wrong property about lists which is **the reverse of a list is always equal to the original**. 
```

Comment:
For this example I'll test an incorrect implementation of a list - **the reverse of a list is always equal to the original**.